pkg/sql/opt/norm/fold_constants_funcs.go

```diff
@@ -321,41 +321,73 @@ func (c *CustomFuncs) FoldUnary(op opt.Operator, input opt.ScalarExpr) (_ opt.Sc
 // by resolving the table name and returning its table ID. This permits the
 // optimizer to do intelligent things like push down filters that look like:
 // ... WHERE oid = 'my_table'::REGCLASS
-func (c *CustomFuncs) foldStringToRegclassCast(
+func (c *CustomFuncs) foldOIDFamilyCast(
 	input opt.ScalarExpr, typ *types.T,
-) (opt.ScalarExpr, error) {
-	// Special case: we're casting a string to a REGCLASS oid, which is a
-	// table id lookup.
+) (_ opt.ScalarExpr, isValid bool, retErr error) {
 	flags := cat.Flags{AvoidDescriptorCaches: false, NoTableStats: true}
 	datum := memo.ExtractConstDatum(input)
-	s := tree.MustBeDString(datum)
-	tn, err := parser.ParseQualifiedTableName(string(s))
-	if err != nil {
-		return nil, err
-	}
-	ds, resName, err := c.f.catalog.ResolveDataSource(c.f.ctx, flags, tn)
-	if err != nil {
-		return nil, err
-	}
 
-	c.mem.Metadata().AddDependency(opt.DepByName(&resName), ds, privilege.SELECT)
+	inputFamily := input.DataType().Family()
+	var dOid *tree.DOid
+	var err error
+
+	switch typ.Oid() {
+	case oid.T_oid:
+		switch inputFamily {
+		case types.StringFamily:
+			s := tree.MustBeDString(datum)
+			dOid, err = tree.ParseDOidAsInt(string(s))
+			if err != nil {
+				return nil, true, err
+			}
+		case types.IntFamily:
+			i := tree.MustBeDInt(datum)
+			dOid, err = tree.IntToOid(i)
+			if err != nil {
+				return nil, true, err
+			}
+		case types.OidFamily:
+			dOid = tree.MustBeDOid(datum)
+		default:
+			return nil, false, nil
+		}
 
-	regclassOid := tree.NewDOidWithName(oid.Oid(ds.PostgresDescriptorID()), types.RegClass, string(tn.ObjectName))
-	return c.f.ConstructConstVal(regclassOid, typ), nil
+	case oid.T_regclass:
+		switch inputFamily {
+		case types.StringFamily:
+			s := tree.MustBeDString(datum)
+			tn, err := parser.ParseQualifiedTableName(string(s))
+			if err != nil {
+				return nil, false, err
+			}
 
+			ds, resName, err := c.f.catalog.ResolveDataSource(c.f.ctx, flags, tn)
+			if err != nil {
+				return nil, false, err
```

Comment:
i guess i was mistaken before, technically the cast is valid, but there was an error, so it should be `return nil, true, err`